### PR TITLE
Optimizations and fixes to write/readBinary()

### PR DESCRIPTION
**Pull request by bradcray**
This optimizes write/readBinary() to use the fast-path "native" code whenever the endianness requested explicitly matches the current architecture's endianness.  Previously, we fell into the slow path for both big/little choices whether or not our architecture is big/little.  Also applied this optimization to readBulkElements() and writeBulkElements().

While testing, I noticed and fixed a bug in which readBinary() was returning the number of bytes read rather than the number of array elements read when using the fast path.  Our only tests of the fast path were testing 8-bit values which is why this was not noticed previously.

While here, I also refactored some logic to exchange multi-case select statements for a singleton call that makes use of a new param->param helper procedure for converting from an endianness enum to an ioKind value.

### CODE REVIEW

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -8936,25 +8956,14 @@ proc fileWriter.writeBinary(const ref data: [?d] ?t, param endian:ioendian = ioe
...
-          }
-          when ioendian.big {
-            e = try _write_binary_internal(this._channel_internal, _iokind.big, b);
-          }
-          when ioendian.little {
-            e = try _write_binary_internal(this._channel_internal, _iokind.little, b);
-          }
-        }
-
-        if e != 0 then break;
```
Comment (benharsh) [inline modules/standard/IO.chpl:8957]: I think this should stick around.

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -8693,6 +8693,26 @@ proc fileWriter.writeBytes(b: bytes, size = b.size) throws {
   try this.writeBinary(b, size);
 }
 
+private proc sysEndianness() {
```
Comment (mppf) [inline modules/standard/IO.chpl:8696]: We could even get this to return a `param`, with some effort on the part of the compiler. I would expect that an approach similar to how we compute that C's `int` is `int(32)` would work (basically, generating some Chapel snippet by looking at the C pre-processor output). `runtime/include/qio/bswap.h` has some related code that might be good to draw from.

Comment (bradcray) [reply modules/standard/IO.chpl:8696]: That felt challenging to do in a cross-compilation environment.  At least, more challenging than I wanted to sign up for here given that it's not called in a critical path (so far, at least).

File: modules/standard/IO.chpl (hunk 1 above)
Comment (bradcray) [reply modules/standard/IO.chpl:8957]: Complete oversight, thanks for catching it.  I'm wondering whether a full paratest would've caught that it was missing as well, and whether I should add a test if not.

**Comment by bradcray** [comment]
@jeremiah-corrado / @benharsh : Would one or both of you have time to give this a review tomorrow?

**Pull request merged**